### PR TITLE
Send notification to refresh user when accept request

### PR DESCRIPTION
**Pull request by Luiz-FS**
**Feature/Bug description:** In certain system actions, such as administration transfer and request to create a link with other institutions, the permissions of the users involved need to be reloaded. Many problems are happening while reloading due to the fact that as soon as notification that a given request has been accepted, the user is reloaded and sometimes there is no time for the queue to add and remove permissions to finish, ie the user is reloaded but does not yet have the permissions. This is because the queue for adding permissions works asynchronously, just like the queue of notifications, so there is currently no way to control the synchronization between the various existing services.

**Solution:** Create a different queue type from the ones that already exist on the system and can be better managed to ensure synchronization of services.
I created a PULL queue where the activities are stored in it and then need to be removed and resolved, I also created a queue manager, it is responsible for creating a notification task, it returns a key that identifies the task newly created, and when it is necessary to solve the task simply pass your key to the manager that it will solve it. In this way, whenever it is necessary to create a notification that needs to be sent soon after some other queues task is resolved, this manager will be able to guarantee the synchronicity between the services.

**Resolved handlers:**
* Institution_parent_request_handler
* Institution_children_request_handler

**TODO/FIXME:** Remove 'if' in the addAdminPermissions queue when all handler notifications that pass through it are resolved.

### CODE REVIEW

File: backend/util/notifications_queue_manager.py [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+import json
+import time
+from google.appengine.api import taskqueue
+from service_messages import send_message_notification
+
+__all__ = ['NotificationsQueueManager', 'Notification', 'NotificationNIL']
+
+queue = taskqueue.Queue('notifications-manage-pull')
+notification_id = {
```
Comment (ruanres) [inline backend/util/notifications_queue_manager.py]: I think that the ids would be more significant if they were the initials of its respective value

File: backend/util/notifications_queue_manager.py [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+class NotificationsQueueManager:
+    """
+    Class to manage the notification queue.
+    """
+    
+    @staticmethod
+    def create_notification_task(notification):
+        """
+        Method for creating a task from the notification and 
+        trigger it in the queue. Returns the task key.
```
Comment (ruanres) [inline backend/util/notifications_queue_manager.py]: ... and add it to the queue

File: backend/util/notifications_queue_manager.py [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        notification_type = notification_id[task_id[:2]]
+        task = find_task(notification_type,task_id)
+
+        if task:
+            notification = create_notification(task_id[:2], **json.loads(task.payload))
+            notification.send_notification()
+            queue.delete_tasks([task])
+
+
+class Notification(object):
```
Comment (ruanres) [inline backend/util/notifications_queue_manager.py]: This class really need to be defined in this file?

File: backend/util/notifications_queue_manager.py [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        return self.__class__.__name__ + " " + str(formated_notification)
+    
+    def __repr__(self):
+        """
+        Method to generate string representation of notification object.
+        """
+        return str(self)
+
+
+class NotificationNIL(Notification):
```
Comment (ruanres) [inline backend/util/notifications_queue_manager.py]: same question

File: backend/util/notifications_queue_manager.py (hunk 1 above)
Comment (Luiz-FS) [reply backend/util/notifications_queue_manager.py]: Placing initials can cause repeated ids.

File: backend/handlers/institution_children_request_handler.py [hunk 5]
```diff
@@ -49,10 +49,13 @@ def put(self, user, request_key):
         institution_children = request.institution_requested_key.get()
         institution_children.set_parent(request.institution_key)
 
-        request.send_response_notification(user.current_institution, user.key, 'ACCEPT')
+        notification_id = request.send_response_notification(user.current_institution, user.key, 'ACCEPT')
```
Comment (mayzabeel) [inline backend/handlers/institution_children_request_handler.py]: the send_response_notification method must be refactored, I think we should separate the creation of the notification and the sending in different methods when it is accepted the invite we call a method that creates and returns the id of the notification and when it is rejected we call send_response_notification

File: backend/models/request_institution_children.py [hunk 6]
```diff
@@ -98,13 +100,24 @@ def send_response_notification(self, current_institution, invitee_key, action):
...
         )
 
-        super(RequestInstitutionChildren, self).send_notification(
-            current_institution=current_institution, 
-            sender_key=invitee_key, 
-            receiver_key=self.sender_key or self.admin_key,
-            notification_type=notification_type,
-            message=notification_message
-        )
+        if action == 'ACCEPT':
```
Comment (mayzabeel) [inline backend/models/request_institution_children.py]: separate creation and submission of notifications in different methods

File: backend/test/model_test/request_institution_children_test.py [hunk 7]
```diff
@@ -213,13 +215,26 @@ def test_send_response_notification(self, create_notification_message, send_noti
...
             notification_type = 'ACCEPT_INSTITUTION_LINK' if action == 'ACCEPT' else 'REJECT_INSTITUTION_LINK'
-            send_notification.assert_called_with(
-                current_institution=self.other_institution.key, 
-                sender_key=self.other_institution.admin, 
-                receiver_key=self.admin.key, 
-                notification_type=notification_type,
-                message='mocked_message'
-            )
+
+            if action == 'ACCEPT':
```
Comment (mayzabeel) [inline backend/test/model_test/request_institution_children_test.py]: separate creation and submission of notifications in different methods

File: frontend/notification/notificationMessageCreatorService.js [hunk 8]
```diff
@@ -44,8 +44,9 @@
             'REJECT_INVITE_USER_ADM': messageCreator('Rejeitou seu convite para ser administrador de ', SINGLE_INST),
             'ACCEPT_INVITE_HIERARCHY': messageCreator('Aceitou seu convite e estabeleceu vínculo entre ', DOUBLE_INST),
             'TRANSFER_ADM_PERMISSIONS': messageCreator('Você agora possui permissões para administrar ', SINGLE_INST),
+            'DELETED_USER': messageCreator('Removeu sua conta na Plataforma Virtual CIS', NO_INST),
+            'ADD_ADM_PERMISSIONS': messageCreator('Suas permissões hierarquicas foram atualizadas na instituição ', SINGLE_INST),
```
Comment (mayzabeel) [inline frontend/notification/notificationMessageCreatorService.js]: hierárquicas

File: backend/worker.py [hunk 9]
```diff
@@ -362,7 +363,13 @@ def addAdminPermissions(self, institution_key):
 
     def post(self):
         institution_key = self.request.get('institution_key')
+        id_nototification = self.request.get('id_notification')
```
Comment (mayzabeel) [inline backend/worker.py]: notification

File: backend/custom_exceptions/queueException.py [hunk 10]
```diff
@@ -0,0 +1,10 @@
+"""Queue Exception."""
+
+__all__ = ['QueueException']
+
+class QueueException(Exception):
+    """Entity Exception."""
```
Comment (mayzabeel) [inline backend/custom_exceptions/queueException.py]: Queue Exception

File: backend/models/request_institution_children.py [hunk 11]
```diff
@@ -87,10 +89,28 @@ def send_notification(self, current_institution):
             notification_type=notification_type,
             message=notification_message
         )
+    
+    def create_accept_response_notification(self, current_institution, invitee_key):
+        """Send accept notification to sender of invite"""
```
Comment (mayzabeel) [inline backend/models/request_institution_children.py]: Create not send

File: backend/util/notification.py [hunk 12]
```diff
@@ -0,0 +1,131 @@
...
+        return self.__class__.__name__ + " " + str(formated_notification)
+    
+    def __repr__(self):
+        """
+        Method to generate string representation of notification object.
+        """
+        return str(self)
+
+
+class NotificationNIL(Notification):
```
Comment (mayzabeel) [inline backend/util/notification.py]: for what?

Comment (Luiz-FS) [reply backend/util/notification.py]: For situations where the same worker may or may not send notifications on certain occasions.

**Pull request merged**